### PR TITLE
Update fit with val hook test

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Related to the Lightning hooks review: https://github.com/PyTorchLightning/pytorch-lightning/issues/7740

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/training_batch_loop.py [hunk 1]
```diff
@@ -342,8 +342,8 @@ def _process_training_step_output(self, training_step_output: STEP_OUTPUT) -> Op
 
         # handle dict return
         if isinstance(training_step_output, dict):
-            loss = training_step_output.pop("loss", None)
-            hiddens = training_step_output.pop("hiddens", None)
+            loss = training_step_output.get("loss")
+            hiddens = training_step_output.get("hiddens")
```
Comment (carmocca) [inline pytorch_lightning/loops/training_batch_loop.py:346]: This silently changed in #7871

The change here restores the previous behaviour. Popping here modifies the output argument for the user too in their `training_step_end`

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 2]
```diff
@@ -366,7 +366,7 @@ def extra(self) -> Dict[str, Any]:
         return self.get('_extra', {})
 
     @extra.setter
-    def extra(self, extra: Mapping[str, Any]) -> None:
+    def extra(self, extra: Dict[str, Any]) -> None:
```
Comment (justusschock) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:369]: why must it be a dict and cannot be any mapping? Is it just because it must be mutable? then we could use `MutableMapping` I think.

File: tests/models/test_hooks.py [hunk 3]
```diff
@@ -381,63 +385,91 @@ def test_trainer_model_hook_system_fit(tmpdir):
...
+        *model._train_batch(trainer, model, train_batches),
+        dict(name='train', args=(False, )),
+        dict(name='on_validation_model_eval'),
+        dict(name='zero_grad'),
+        dict(name='Callback.on_validation_start', args=(trainer, model)),
+        dict(name='on_validation_start'),
+        *model._eval_epoch('validation', trainer, model, val_batches, 'x'),
+        dict(name='Callback.on_validation_end', args=(trainer, model)),
+        # `ModelCheckpoint.save_checkpoint` is called here from `Callback.on_validation_end`
+        dict(name='Callback.on_save_checkpoint', args=(trainer, model)),
```
Comment (awaelchli) [inline tests/models/test_hooks.py:445]: this also has a checkpoint arg.
you made it so that missing args don't fail if omitted?

File: tests/models/test_hooks.py [hunk 4]
```diff
@@ -283,35 +283,38 @@ def test_epoch_end(self, *args, **kwargs):
...
-            'validation_step',
-            'validation_step_end',
-            'on_validation_batch_end',
-        ]
+    def _train_batch(trainer, model, batches):
+        out = []
+        for i in range(batches):
+            out.extend([
+                # TODO: `{,Callback}.on_batch_{start,end}`
+                dict(name='Callback.on_batch_start', args=(trainer, model)),
```
Comment (justusschock) [inline tests/models/test_hooks.py:291]: is the explicit constructor here (instead of `{ }`) due to yapf formatting or for another reason?

File: tests/models/test_hooks.py [hunk 5]
```diff
@@ -283,35 +283,38 @@ def test_epoch_end(self, *args, **kwargs):
...
-            'forward',
-            'validation_step',
-            'validation_step_end',
-            'on_validation_batch_end',
-        ]
+    def _train_batch(trainer, model, batches):
+        out = []
+        for i in range(batches):
+            out.extend([
+                # TODO: `{,Callback}.on_batch_{start,end}`
```
Comment (justusschock) [inline tests/models/test_hooks.py:290]: Why is this still a todo? Are there issues here?

File: tests/models/test_hooks.py [hunk 6]
```diff
@@ -283,35 +283,38 @@ def test_epoch_end(self, *args, **kwargs):
...
+                dict(name='training_step', args=(ANY, i)),
+                dict(name='training_step_end', args=(dict(loss=ANY), )),
+                dict(name='Callback.on_before_zero_grad', args=(trainer, model, ANY)),
+                dict(name='on_before_zero_grad', args=(ANY, )),
+                dict(name='optimizer_zero_grad', args=(0, i, ANY, 0)),
+                # TODO: `on_before_backward`
+                dict(name='backward', args=(ANY, ANY, 0)),
+                dict(name='Callback.on_after_backward', args=(trainer, model)),
+                dict(name='on_after_backward'),
+                # TODO: `on_before_optimizer_step`
```
Comment (justusschock) [inline tests/models/test_hooks.py:307]: Same as above (issues?)

File: tests/models/test_hooks.py [hunk 7]
```diff
@@ -283,35 +283,38 @@ def test_epoch_end(self, *args, **kwargs):
...
+                dict(name='on_before_batch_transfer', args=(ANY, None)),
+                dict(name='transfer_batch_to_device', args=(ANY, torch.device('cpu'), None)),
+                dict(name='on_after_batch_transfer', args=(ANY, None)),
+                dict(name='forward', args=(ANY, )),
+                dict(name='training_step', args=(ANY, i)),
+                dict(name='training_step_end', args=(dict(loss=ANY), )),
+                dict(name='Callback.on_before_zero_grad', args=(trainer, model, ANY)),
+                dict(name='on_before_zero_grad', args=(ANY, )),
+                dict(name='optimizer_zero_grad', args=(0, i, ANY, 0)),
+                # TODO: `on_before_backward`
```
Comment (justusschock) [inline tests/models/test_hooks.py:303]: Same as above (issues?)

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:369]: > Is it just because it must be mutable?

Yes

>  then we could use MutableMapping I think.

Sure, I can change it. What's the difference?

File: tests/models/test_hooks.py (hunk 4 above)
Comment (carmocca) [reply tests/models/test_hooks.py:291]: yapf formatting :(

File: tests/models/test_hooks.py (hunk 5 above)
Comment (carmocca) [reply tests/models/test_hooks.py:290]: I'm using TODOs in this file to point out inconsistencies between functions but not planning to resolve them here.

Comment (carmocca) [reply tests/models/test_hooks.py:290]: Further discussions about these can be had  in #7740

File: tests/models/test_hooks.py (hunk 6 above)
Comment (carmocca) [reply tests/models/test_hooks.py:307]: Same answer :)

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:369]: I think MutableMapping is not as restrictive ( a dict is a special case of a muable mapping). But I'm fine with Dict as well (in fact that's the only mutable mapping people commonly use in practice, I think).

File: tests/models/test_hooks.py (hunk 5 above)
Comment (carmocca) [reply tests/models/test_hooks.py:290]:
```suggestion
                # TODO: `.on_batch_{start,end}`
```

**Review by Borda — APPROVED** [review]
